File: python/networking/ftp_program.py
"""FTP Program — file transfer protocol client/server model."""


class FtpServer:
    """A simplified FTP server with a virtual filesystem."""

    def __init__(self):
        self._files: dict[str, str] = {}

    def add_file(self, name: str, content: str) -> None:
        """Seed the virtual filesystem with a file."""
        self._files[name] = content

    def handle(self, command: str) -> str:
        """Process a command string and return a response string."""
        parts = command.strip().split(None, 2)
        if not parts:
            return "500 Syntax error"

        cmd = parts[0].upper()

        if cmd == "LIST":
            if not self._files:
                return "226 (empty)"
            listing = "\n".join(sorted(self._files.keys()))
            return f"226 File list:\n{listing}"

        elif cmd == "RETR":
            if len(parts) < 2:
                return "501 Syntax error in parameters"
            filename = parts[1]
            if filename not in self._files:
                return f"550 {filename}: No such file"
            return f"226 Transfer complete:\n{self._files[filename]}"

        elif cmd == "STOR":
            if len(parts) < 3:
                return "501 Syntax error in parameters"
            filename = parts[1]
            content = parts[2]
            self._files[filename] = content
            return f"226 {filename} stored ({len(content)} bytes)"

        elif cmd == "DELE":
            if len(parts) < 2:
                return "501 Syntax error in parameters"
            filename = parts[1]
            if filename not in self._files:
                return f"550 {filename}: No such file"
            del self._files[filename]
            return f"250 {filename} deleted"

        elif cmd == "SIZE":
            if len(parts) < 2:
                return "501 Syntax error in parameters"
            filename = parts[1]
            if filename not in self._files:
                return f"550 {filename}: No such file"
            return f"213 {len(self._files[filename])}"

        elif cmd == "QUIT":
            return "221 Goodbye"

        else:
            return f"502 Command not implemented: {cmd}"


class FtpClient:
    """A simplified FTP client that sends commands to an FtpServer."""

    def __init__(self):
        self._server: FtpServer | None = None
        self._connected = False

    def connect(self, server: FtpServer) -> str:
        """Connect to an FTP server instance."""
        self._server = server
        self._connected = True
        return "220 Service ready"

    def send(self, command: str) -> str:
        """Send a command to the connected server and return the response."""
        if not self._connected or self._server is None:
            return "421 Not connected"
        response = self._server.handle(command)
        if command.strip().upper() == "QUIT":
            self._connected = False
            self._server = None
        return response

    def list_files(self) -> str:
        return self.send("LIST")

    def get(self, filename: str) -> str:
        return self.send(f"RETR {filename}")

    def put(self, filename: str, content: str) -> str:
        return self.send(f"STOR {filename} {content}")

    def delete(self, filename: str) -> str:
        return self.send(f"DELE {filename}")

    def size(self, filename: str) -> str:
        return self.send(f"SIZE {filename}")

    def quit(self) -> str:
        return self.send("QUIT")


if __name__ == "__main__":
    # Set up server with initial files
    server = FtpServer()
    server.add_file("readme.txt", "Welcome to the FTP server.")
    server.add_file("data.csv", "name,value\nalpha,1\nbeta,2")

    # Connect client
    client = FtpClient()
    print(client.connect(server))

    # List files
    print("\n--- LIST ---")
    print(client.list_files())

    # Download a file
    print("\n--- RETR readme.txt ---")
    print(client.get("readme.txt"))

    # Upload a new file
    print("\n--- STOR notes.txt ---")
    print(client.put("notes.txt", "These are my notes."))

    # List again to see the new file
    print("\n--- LIST ---")
    print(client.list_files())

    # Check file size
    print("\n--- SIZE data.csv ---")
    print(client.size("data.csv"))

    # Delete a file
    print("\n--- DELE data.csv ---")
    print(client.delete("data.csv"))

    # List after deletion
    print("\n--- LIST ---")
    print(client.list_files())

    # Quit
    print("\n--- QUIT ---")
    print(client.quit())

    # Try to send after quit
    print("\n--- After disconnect ---")
    print(client.list_files())

File: rust/networking/ftp_program.rs
//! FTP Program — file transfer protocol client/server model.

use std::collections::BTreeMap;

/// A simplified FTP server with a virtual filesystem.
pub struct FtpServer {
    files: BTreeMap<String, String>,
}

impl FtpServer {
    pub fn new() -> Self {
        FtpServer {
            files: BTreeMap::new(),
        }
    }

    /// Seed the virtual filesystem with a file.
    pub fn add_file(&mut self, name: &str, content: &str) {
        self.files.insert(name.to_string(), content.to_string());
    }

    /// Process a command string and return a response string.
    pub fn handle(&mut self, command: &str) -> String {
        let trimmed = command.trim();
        let mut parts = trimmed.splitn(3, ' ');
        let cmd = match parts.next() {
            Some(c) => c.to_uppercase(),
            None => return "500 Syntax error".to_string(),
        };

        match cmd.as_str() {
            "LIST" => {
                if self.files.is_empty() {
                    "226 (empty)".to_string()
                } else {
                    let listing: Vec<&str> = self.files.keys().map(|s| s.as_str()).collect();
                    format!("226 File list:\n{}", listing.join("\n"))
                }
            }
            "RETR" => {
                let filename = match parts.next() {
                    Some(f) => f,
                    None => return "501 Syntax error in parameters".to_string(),
                };
                match self.files.get(filename) {
                    Some(content) => format!("226 Transfer complete:\n{}", content),
                    None => format!("550 {}: No such file", filename),
                }
            }
            "STOR" => {
                let filename = match parts.next() {
                    Some(f) => f,
                    None => return "501 Syntax error in parameters".to_string(),
                };
                let content = match parts.next() {
                    Some(c) => c,
                    None => return "501 Syntax error in parameters".to_string(),
                };
                let len = content.len();
                self.files.insert(filename.to_string(), content.to_string());
                format!("226 {} stored ({} bytes)", filename, len)
            }
            "DELE" => {
                let filename = match parts.next() {
                    Some(f) => f,
                    None => return "501 Syntax error in parameters".to_string(),
                };
                if self.files.remove(filename).is_some() {
                    format!("250 {} deleted", filename)
                } else {
                    format!("550 {}: No such file", filename)
                }
            }
            "SIZE" => {
                let filename = match parts.next() {
                    Some(f) => f,
                    None => return "501 Syntax error in parameters".to_string(),
                };
                match self.files.get(filename) {
                    Some(content) => format!("213 {}", content.len()),
                    None => format!("550 {}: No such file", filename),
                }
            }
            "QUIT" => "221 Goodbye".to_string(),
            _ => format!("502 Command not implemented: {}", cmd),
        }
    }
}

/// A simplified FTP client that sends commands to an FtpServer.
pub struct FtpClient<'a> {
    server: Option<&'a mut FtpServer>,
}

impl<'a> FtpClient<'a> {
    pub fn new() -> Self {
        FtpClient { server: None }
    }

    pub fn connect(&mut self, server: &'a mut FtpServer) -> String {
        self.server = Some(server);
        "220 Service ready".to_string()
    }

    pub fn send(&mut self, command: &str) -> String {
        let server = match self.server.as_mut() {
            Some(s) => s,
            None => return "421 Not connected".to_string(),
        };
        let response = server.handle(command);
        if command.trim().eq_ignore_ascii_case("QUIT") {
            self.server = None;
        }
        response
    }

    pub fn list_files(&mut self) -> String {
        self.send("LIST")
    }

    pub fn get(&mut self, filename: &str) -> String {
        self.send(&format!("RETR {}", filename))
    }

    pub fn put(&mut self, filename: &str, content: &str) -> String {
        self.send(&format!("STOR {} {}", filename, content))
    }

    pub fn delete(&mut self, filename: &str) -> String {
        self.send(&format!("DELE {}", filename))
    }

    pub fn size(&mut self, filename: &str) -> String {
        self.send(&format!("SIZE {}", filename))
    }

    pub fn quit(&mut self) -> String {
        self.send("QUIT")
    }
}

fn main() {
    let mut server = FtpServer::new();
    server.add_file("readme.txt", "Welcome to the FTP server.");
    server.add_file("data.csv", "name,value\nalpha,1\nbeta,2");

    let mut client = FtpClient::new();
    println!("{}", client.connect(&mut server));

    println!("\n--- LIST ---");
    println!("{}", client.list_files());

    println!("\n--- RETR readme.txt ---");
    println!("{}", client.get("readme.txt"));

    println!("\n--- STOR notes.txt ---");
    println!("{}", client.put("notes.txt", "These are my notes."));

    println!("\n--- LIST ---");
    println!("{}", client.list_files());

    println!("\n--- SIZE data.csv ---");
    println!("{}", client.size("data.csv"));

    println!("\n--- DELE data.csv ---");
    println!("{}", client.delete("data.csv"));

    println!("\n--- QUIT ---");
    println!("{}", client.quit());
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_list_empty() {
        let mut server = FtpServer::new();
        assert_eq!(server.handle("LIST"), "226 (empty)");
    }

    #[test]
    fn test_list_with_files() {
        let mut server = FtpServer::new();
        server.add_file("b.txt", "bee");
        server.add_file("a.txt", "ay");
        let response = server.handle("LIST");
        assert!(response.starts_with("226 File list:"));
        assert!(response.contains("a.txt"));
        assert!(response.contains("b.txt"));
    }

    #[test]
    fn test_stor_and_retr() {
        let mut server = FtpServer::new();
        let store_resp = server.handle("STOR hello.txt Hello world");
        assert!(store_resp.starts_with("226"));
        let retr_resp = server.handle("RETR hello.txt");
        assert!(retr_resp.contains("Hello world"));
    }

    #[test]
    fn test_retr_missing() {
        let mut server = FtpServer::new();
        let resp = server.handle("RETR ghost.txt");
        assert!(resp.starts_with("550"));
    }

    #[test]
    fn test_dele() {
        let mut server = FtpServer::new();
        server.add_file("temp.txt", "data");
        let resp = server.handle("DELE temp.txt");
        assert!(resp.starts_with("250"));
        let resp2 = server.handle("RETR temp.txt");
        assert!(resp2.starts_with("550"));
    }

    #[test]
    fn test_dele_missing() {
        let mut server = FtpServer::new();
        let resp = server.handle("DELE nope.txt");
        assert!(resp.starts_with("550"));
    }

    #[test]
    fn test_size() {
        let mut server = FtpServer::new();
        server.add_file("f.txt", "12345");
        assert_eq!(server.handle("SIZE f.txt"), "213 5");
    }

    #[test]
    fn test_quit() {
        let mut server = FtpServer::new();
        assert_eq!(server.handle("QUIT"), "221 Goodbye");
    }

    #[test]
    fn test_unknown_command() {
        let mut server = FtpServer::new();
        let resp = server.handle("NOOP");
        assert!(resp.starts_with("502"));
    }

    #[test]
    fn test_client_not_connected() {
        let mut client = FtpClient::new();
        assert_eq!(client.list_files(), "421 Not connected");
    }

    #[test]
    fn test_client_disconnect_on_quit() {
        let mut server = FtpServer::new();
        let mut client = FtpClient::new();
        client.connect(&mut server);
        client.quit();
        assert_eq!(client.list_files(), "421 Not connected");
    }
}
